src/privacy_accounting/optimizer/dp_optimizer.py

```diff
@@ -0,0 +1,100 @@
+from math import ceil
+
+import torch
+
+import src.utils.torch_nest_utils as nest
+from src.privacy_accounting.analysis import QueryWithLedger
+from src.privacy_accounting.dp_query import GaussianDPQuery
+from . import WrapperOptimiser
+
+
+class DPOptimiser(WrapperOptimiser):
+
+    def __init__(self,
+                 optimiser,
+                 model,
+                 vector_loss,
+                 dp_sum_query,
+                 num_microbatches=None):
+
+        self.optimiser = optimiser
+        self.model = model
+        self.vector_loss = vector_loss
+        self.dp_sum_query = dp_sum_query
+        self.num_microbatches = num_microbatches
+
+        self._global_parameters = self.dp_sum_query.initial_global_state()
+
+    def fit_batch(self, x: torch.Tensor, y: torch.Tensor):
+
+        loss = self.vector_loss(self.model(x), y)
+
+        param_groups = self.optimiser.param_groups
+
+        sample_state = self.dp_sum_query.initial_sample_state(
+            nest.parameters_to_tensor_groups(param_groups, 'data')
+        )
+        sample_params = self.dp_sum_query.derive_sample_params(self._global_parameters)
+
+        microbatch_size = 1 if self.num_microbatches is None else ceil(loss.shape[0] / self.num_microbatches)
+
+        microbatches_losses = loss.split(microbatch_size, dim=0)
+
+        def process_microbatch(losses, sample_state):
+            self.optimiser.zero_grad()
+            microbatch_loss = losses.mean(dim=0)
```

Comment:
Mean on the gradients contributing to a micorbatch I'm p sure, sum on the clipped microbatch gradients.

Reply:
This is the proceduce laid out in https://arxiv.org/pdf/1812.06210.pdf for micro-batching, I think it is correct. We use micro-batching to create better gradient estimates before clipping, hence we average rather than sum. When setting learning rates, this could be something to bear in mind youre right.